gnovm/pkg/gnolang/store.go

```diff
@@ -73,53 +73,276 @@ type Store interface {
+type hashMap[K comparable, V any] interface {
+	Get(K) (V, bool)
+	Set(K, V)
+	Delete(K)
+	Iterate() func(yield func(K, V) bool)
+}
+
+type txLogMap[K comparable, V any] struct {
+	source hashMap[K, V]
+	dirty  map[K]deletable[V]
+}
+
+func newTxLog[K comparable, V any](source hashMap[K, V]) *txLogMap[K, V] {
+	return &txLogMap[K, V]{
+		source: source,
+		dirty:  make(map[K]deletable[V]),
+	}
+}
+
+// write commits the data in dirty to the map in source.
+func (b *txLogMap[K, V]) write() {
+	for k, v := range b.dirty {
+		if v.deleted {
+			b.source.Delete(k)
+		} else {
+			b.source.Set(k, v.v)
+		}
+	}
+	b.dirty = make(map[K]deletable[V])
+}
+
+func (b txLogMap[K, V]) Get(k K) (V, bool) {
+	if bufValue, ok := b.dirty[k]; ok {
+		if bufValue.deleted {
+			var zeroV V
+			return zeroV, false
+		}
+		return bufValue.v, true
+	}
+
+	return b.source.Get(k)
+}
+
+func (b txLogMap[K, V]) Set(k K, v V) {
+	b.dirty[k] = deletable[V]{v: v}
+}
+
+func (b txLogMap[K, V]) Delete(k K) {
+	b.dirty[k] = deletable[V]{deleted: true}
+}
+
+func (b txLogMap[K, V]) Iterate() func(yield func(K, V) bool) {
+	return func(yield func(K, V) bool) {
+		b.source.Iterate()(func(k K, v V) bool {
+			if dirty, ok := b.dirty[k]; ok {
+				if dirty.deleted {
+					return true
+				}
+				return yield(k, dirty.v)
+			}
+
+			// not in dirty
+			return yield(k, v)
+		})
+		// yield for new values
+		for k, v := range b.dirty {
+			if v.deleted {
+				continue
+			}
+			_, ok := b.source.Get(k)
+			if ok {
+				continue
+			}
+			if !yield(k, v.v) {
+				break
+			}
+		}
+	}
+}
+
+type mapWrapper[K comparable, V any] map[K]V
+
+func (m mapWrapper[K, V]) Get(k K) (V, bool) {
+	v, ok := m[k]
+	return v, ok
+}
+
+func (m mapWrapper[K, V]) Set(k K, v V) {
+	m[k] = v
+}
+
+func (m mapWrapper[K, V]) Delete(k K) {
+	delete(m, k)
+}
+
+func (m mapWrapper[K, V]) Iterate() func(yield func(K, V) bool) {
+	return func(yield func(K, V) bool) {
+		for k, v := range m {
+			if !yield(k, v) {
+				return
+			}
+		}
+	}
+}
+
+type deletable[V any] struct {
+	v       V
+	deleted bool
 }
 
 func NewStore(alloc *Allocator, baseStore, iavlStore store.Store) *defaultStore {
 	ds := &defaultStore{
-		alloc:            alloc,
+		baseStore: baseStore,
+		iavlStore: iavlStore,
+		alloc:     alloc,
+
+		// cacheObjects is set; objects in the store will be copied over for any transaction.
+		cacheObjects: make(map[ObjectID]Object),
+		cacheTypes:   mapWrapper[TypeID, Type](map[TypeID]Type{}),
+		cacheNodes:   mapWrapper[Location, BlockNode](map[Location]BlockNode{}),
+
+		// store configuration
 		pkgGetter:        nil,
-		cacheObjects:     make(map[ObjectID]Object),
-		cacheTypes:       make(map[TypeID]Type),
-		cacheNodes:       make(map[Location]BlockNode),
 		cacheNativeTypes: make(map[reflect.Type]Type),
-		baseStore:        baseStore,
-		iavlStore:        iavlStore,
+		pkgInjector:      nil,
+		nativeStore:      nil,
 		go2gnoStrict:     true,
 	}
 	InitStoreCaches(ds)
 	return ds
 }
 
+// Unstable.
```

Comment:
What does this mean?

Reply:
Beats me.

This code has been moved (due to having deleted it and adding it again); likely originally written again.

I'll move it back to its original place, for diff's sake.